lib/services/s3.js

```diff
@@ -240,7 +240,7 @@ AWS.util.update(AWS.S3.prototype, {
     var bucket = httpRequest.virtualHostedBucket;
     if (bucket && httpRequest.path) {
       if (req.params && req.params.Key) {
-        if (httpRequest.path.indexOf('/' + AWS.util.uriEscapePath(req.params.Key)) === 0) {
+        if (httpRequest.path.indexOf('/' + AWS.util.uriEscapePath(req.params.Key)) === 0 && httpRequest.endpoint.host.indexOf(req.service.endpoint.host) === -1) {
```

Comment:
Are you trying to make sure Bucket is ready moved from path to host? This statement fixes the issue but I'm wondering what will happen if the request is using [dualstack endpoints](https://github.com/aws/aws-sdk-js/blob/master/lib/region_config.js#L53), will the appended statement always true?

I think you can be more specific in the if condition. For example you can check if `Key` parameter exists in the request path **and** `Bucket` parameter not in the path **and [optional]** `Bucket` parameter is appended to host